lib/vm/src/instance/mod.rs

```diff
@@ -776,6 +782,205 @@ impl Instance {
+        memory_index: LocalMemoryIndex,
+        dst: u32,
+        val: u64,
+        timeout: i64,
+    ) -> Result<u32, Trap> {
+        let memory = self.memory(memory_index);
+        //if ! memory.shared {
+        // We should trap according to spec, but official test rely on not trapping...
+        //}
+
+        let ret = unsafe { memory32_atomic_check64(&memory, dst, val) };
+
+        if let Ok(mut ret) = ret {
+            if ret == 0 {
+                ret = self.do_wait(memory_index.as_u32(), dst, timeout);
+            }
+            Ok(ret)
+        } else {
+            ret
+        }
+    }
+
+    /// Perform an Atomic.Wait64
+    pub(crate) fn imported_memory_wait64(
+        &mut self,
+        memory_index: MemoryIndex,
+        dst: u32,
+        val: u64,
+        timeout: i64,
+    ) -> Result<u32, Trap> {
+        let import = self.imported_memory(memory_index);
+        let memory = unsafe { import.definition.as_ref() };
+        //if ! memory.shared {
+        // We should trap according to spec, but official test rely on not trapping...
+        //}
+
+        let ret = unsafe { memory32_atomic_check64(memory, dst, val) };
+
+        if let Ok(mut ret) = ret {
+            if ret == 0 {
+                ret = self.do_wait(memory_index.as_u32(), dst, timeout);
+            }
+            Ok(ret)
+        } else {
+            ret
+        }
+    }
+
+    /// Perform an Atomic.Notify
+    pub(crate) fn local_memory_notify(
+        &mut self,
+        memory_index: LocalMemoryIndex,
+        dst: u32,
+        count: u32,
+    ) -> Result<u32, Trap> {
+        //let memory = self.memory(memory_index);
+        //if ! memory.shared {
+        // We should trap according to spec, but official test rely on not trapping...
+        //}
+
+        // fetch the notifier
+        let key = (memory_index.as_u32(), dst);
+        let mut conds = self.conditions.lock().unwrap();
+        let mut cnt = 0u32;
+        if conds.contains_key(&key) {
+            let v = conds.get_mut(&key).unwrap();
+            for (t, b) in v {
+                if cnt < count {
+                    *b = true; // mark as was waiked up
+                    t.unpark(); // wakeup!
+                    cnt += 1;
+                }
+            }
+        }
+        Ok(cnt)
+    }
+    /// Perform an Atomic.Notify
+    pub(crate) fn imported_memory_notify(
+        &mut self,
+        memory_index: MemoryIndex,
+        dst: u32,
+        count: u32,
+    ) -> Result<u32, Trap> {
+        //let import = self.imported_memory(memory_index);
+        //let memory = unsafe { import.definition.as_ref() };
+        //if ! memory.shared {
+        // We should trap according to spec, but official test rely on not trapping...
+        //}
+
+        // fetch the notifier
+        let key = (memory_index.as_u32(), dst);
+        let mut conds = self.conditions.lock().unwrap();
+        let mut cnt = 0u32;
+        if conds.contains_key(&key) {
```

Comment:
This is the same code as above, right?
I'd put it in a separate function to avoid duplication.

Also: double hashing/acquiring ( as above).

Reply:
Not sure, I'll check if it's not a borrow checker hell.